api/qrcodegenerator.proto

```diff
@@ -0,0 +1,108 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+import "github.com/envoyproxy/protoc-gen-validate/validate/validate.proto";
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+import "google/api/annotations.proto";
+import "google/protobuf/empty.proto";
+import "google/protobuf/field_mask.proto";
+import "lorawan-stack/api/end_device.proto";
+
+package ttn.lorawan.v3;
+
+option go_package = "go.thethings.network/lorawan-stack/pkg/ttnpb";
+
+message QRCodeFormat {
+  option (gogoproto.populate) = false;
+
+  string name = 1 [(validate.rules).string.max_len = 100];
+  string description = 2 [(validate.rules).string.max_len = 200];
+  // The entity fields required to generate the QR code.
+  google.protobuf.FieldMask field_mask = 3 [(gogoproto.nullable) = false];
+}
+
+message QRCodeFormats {
+  option (gogoproto.populate) = false;
+
+  map<string, QRCodeFormat> formats = 1 [(validate.rules).map.keys.string = {pattern: "^[a-z0-9](?:[-]?[a-z0-9]){2,}$", max_len: 36}];
+}
+
+message GetQRCodeFormatRequest {
+  option (gogoproto.populate) = false;
+
+  string format_id = 1 [(gogoproto.customname) = "FormatID", (validate.rules).string = {pattern: "^[a-z0-9](?:[-]?[a-z0-9]){2,}$", max_len: 36}];
+}
+
+message GenerateEndDeviceQRCodeTextRequest {
+  option (gogoproto.populate) = false;
+
+  string format_id = 1 [(gogoproto.customname) = "FormatID", (validate.rules).string = {pattern: "^[a-z0-9](?:[-]?[a-z0-9]){2,}$", max_len: 36}];
+  EndDevice end_device = 2 [(gogoproto.nullable) = false, (validate.rules).message.required = true];
+}
+
+message GenerateQRCodeTextResponse {
+  option (gogoproto.populate) = false;
+
+  string text = 1;
+}
+
+message GenerateEndDeviceQRCodeImageRequest {
+  option (gogoproto.populate) = false;
+
+  string format_id = 1 [(gogoproto.customname) = "FormatID", (validate.rules).string = {pattern: "^[a-z0-9](?:[-]?[a-z0-9]){2,}$", max_len: 36}];
+  EndDevice end_device = 2 [(gogoproto.nullable) = false, (validate.rules).message.required = true];
+  uint32 image_size = 3 [(validate.rules).uint32.gte = 10, (validate.rules).uint32.lte = 1000];
+}
+
+message GenerateQRCodeImageResponse {
+  option (gogoproto.populate) = false;
+
+  string text = 1;
+  // QR code in PNG format.
+  bytes image = 2;
+}
+
+service EndDeviceQRCodeGenerator {
+  // Return the QR code format.
+  rpc GetFormat(GetQRCodeFormatRequest) returns (QRCodeFormat) {
+    option (google.api.http) = {
+      get: "/qrcodes/end-devices/formats/{format_id}"
+    };
+  };
+
+  // Returns the supported formats.
+  rpc ListFormats(google.protobuf.Empty) returns (QRCodeFormats) {
+    option (google.api.http) = {
+      get: "/qrcodes/end-devices/formats"
+    };
+  };
+
+  // Generates the text value of the QR code.
+  rpc GenerateText(GenerateEndDeviceQRCodeTextRequest) returns (GenerateQRCodeTextResponse) {
```

Comment:
Since `GenerateImage` also includes `text`, isn't it better to merge `GenerateText` and `GenerateImage` into one RPC and have a request message like

```proto
message GenerateEndDeviceQRCodeRequest {
  string format_id = 1 [(gogoproto.customname) = "FormatID", (validate.rules).string = {pattern: "^[a-z0-9](?:[-]?[a-z0-9]){2,}$", max_len: 36}];
  EndDevice end_device = 2 [(gogoproto.nullable) = false, (validate.rules).message.required = true];

  message ImageConfig {
    (maybe) image_type = 1; // bmp, png, ...
    uint32 image_size = 2 [(validate.rules).uint32.gte = 10, (validate.rules).uint32.lte = 1000];
  }

  ImageConfig image_config = 3; // if set, include the QR code rendered as an image
}
```